events.js

```diff
@@ -1,5 +1,37 @@
 "use strict"
 
+// correctly handle back/forward buttons
+
+var mouseOnPage = true
+document.addEventListener("DOMContentLoaded", function() {
+    document.body.addEventListener("mouseenter", function() {
+        mouseOnPage = true
+    });
+    document.body.addEventListener("mouseleave", function() {
+        mouseOnPage = false
+    });
+});
+window.addEventListener("hashchange", function() {
```

Comment:
On consideration, "hashchange" is probably the wrong event. This business with "keepHash" and mouseOnPage is fiddly, and it fails when using keyboard shortcuts rather than clicking on the buttons.

The "popstate" event seems like a better fit. It should be less error-prone and should eliminate a lot of the fiddly bits of this code.

I'm also inclined to think that this event handler should only be installed after the dataset is loaded for the first time, though in practice you'd have to be trying pretty hard to make a problem out of this.

Reply:
popstate is also called when updating the hash, so it would act almost exactly the same as hashchange. How do you think popstate would eliminate the keepHash (I think keepHash is ugly - I don't like it either).

> I'm also inclined to think that this event handler should only be installed after the dataset is loaded for the first time, though in practice you'd have to be trying pretty hard to make a problem out of this.

I don't think it would be possible to make a problem out of this unless you automated something specifically to break it - and even then it would only affect a local javascript that would be fixed on reload.

Reply:
Bluh, for some reason I thought popstate wouldn't do that, but okay. hashchange it is.

A workaround, then, would be to define a function that removes the event handler, changes the hash, and then re-adds the event handler. Then this function could be used in the handful of places where it assigns to window.location.hash. This would ensure that the only things that can trigger the event aren't other events.

Reply:
Instead of removing the handler there could be a check in the handler for a global flag that is set when the handler is triggered and removed when the handler is finished.

i.e. 
```
var navigationInProgress = false
window.addEventListener("hashChange", function() {
    if (navigationInProgress) {
        return
    }
    navigationInProgress = true
    // do stuff
    navigationInProgress = false
}
```

Reply:
This is probably cleaner, sure.